File: p5js/GridJoe/js/target.js
"use strict";

class Target {

    constructor() {
        this.speedMin = 0.01;  // float
        this.speedMax = 0.05;  // float
        this.speed;  // float
        this.clickOdds = 0.1;  // float
        this.chooseOdds = 0.01;  // float
        this.markTime = 0;  // int
        this.timeInterval = 200;  // int
    
        this.posX = width/2;  // float
        this.posY = height/2;  // float
        this.targetX;  // float
        this.targetY;  // float
        this.minDist = 5;  // int
        this.clicked = false;
        this.armResetAll = false;
        
        this.pickTarget();
    }

    run() {
        this.posX = lerp(this.posX, this.targetX, this.speed);
        this.posY = lerp(this.posY, this.targetY, this.speed);
        
        if (millis() > this.markTime + this.timeInterval || dist(this.posX, this.posY, this.targetX, this.targetY) < this.minDist) {
            this.pickTarget();
        }
    }
    
    pickTarget() {
        this.markTime = millis();
        
        this.targetX = lerp(this.posX, random(0, width), 0.5);
        this.targetY = lerp(this.posY, random(0, height), 0.5);
        
        this.speed = random(this.speedMin, this.speedMax);
        let r = random(1);
        if (r < this.clickOdds) this.clicked = !this.clicked;
        if (r < this.chooseOdds) this.armResetAll = true;
    }

}

File: openframeworks/GridJoeSyphon/src/Target.cpp
#include "Target.h"
#include "ofApp.h"

Target::Target(int w, int h) {
    sW = w;
    sH = h;
    
    speedMin = 0.01;
    speedMax = 0.05;
    clickOdds = 0.1;
    chooseOdds = 0.01;
    markTime = 0;
    timeInterval = 200;

    posX = sW/2;
    posY = sH/2;
    minDist = 5;
    clicked = false;
    armResetAll = false;
    
    pickTarget();
}

void Target::run() {
    posX = ofLerp(posX, targetX, speed);
    posY = ofLerp(posY, targetY, speed);

    if (ofGetElapsedTimeMillis() > markTime + timeInterval || ofDist(posX, posY, targetX, targetY) < minDist) {
      pickTarget();
    }
}

void Target::pickTarget() {
    markTime = ofGetElapsedTimeMillis();

    targetX = ofLerp(posX, ofRandom(0, sW), 0.5);
    targetY = ofLerp(posY, ofRandom(0, sH), 0.5);

    speed = ofRandom(speedMin, speedMax);
    float r = ofRandom(1);
    if (r < clickOdds) clicked = !clicked;
    if (r < chooseOdds) armResetAll = true;
}
  
